apluslms_roman/observer.py

```diff
@@ -151,38 +153,81 @@ def container_msg(self, step, msg):
         self._send_message(Message.CONTAINER_MSG, step, msg)
 
     def result_msg(self, result):
+        state = (StepState.SUCCEEDED if result.ok
+            else (StepState.CANCELLED if result.cancelled
+                else StepState.FAILED))
+        self._state_update(result.step, state)
```

Comment:
I was thinking something like this:

```
def result_msg(self, result):
        state = (StepState.SUCCEEDED if result.ok else
                 StepState.CANCELLED if result.cancelled else
                 StepState.FAILED)
        steps = (result.step,) if result.step else self._states
        for step in steps:
            self._state_update(step, state)
        self._message(self._phase, Message.RESULT_MSG, result.step, state, (result.code, result.error))
```
Presume that for `SUCCEEDED`, there is no step given (as all steps must succeed for result to be ok). Thus, what we can do (but shouldn't be mandatory), is to ensure all steps are marked (and notified) to be in `SUCCEEDED` state. After that, we will send result msg (directly as state would be `UNKNOWN` for None state).

This should make it possible for result message to have optional step.

Reply:
Related to another comment, `_state_update` test should probable be change from `if self.get_step_state(step) != state:` to `cur_state = self.get_step_state(step); if cur_state != state and not cur_state.completed:`.

Reply:
...so should all of the steps' states be changed to `SUCCEEDED`? or can we just ignore the steps and send a result message?

Reply:
.completed is different than `SUCCEEDED`. It's any state that means step has run to the end (SUCCEEDED, FAILED, CANCELLED). Does this answer the question?

Reply:
> Thus, what we can do (but shouldn't be mandatory), is to ensure all steps are marked (and notified) to be in SUCCEEDED state.
I was referring to that

Reply:
If we end up with ok result, that means all the steps must have succeeded, thus they can be marked as such (and notified).

Reply:
With the current system, if an earlier step for some reason hasn't been marked as succeeded, the observer will print out `ok: [time]`, without really telling which step is in question (and the time will be inaccurate, too), should that be changed?

Reply:
Add a comment about that, but it's fine as it normally doesn't happen (requires a bug). I have a plan to add serialization to streamobserver, which would actually take care of this.